core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -126,26 +126,136 @@ class FileMessageSet private[kafka](@volatile var file: File,
+    catch {
+      case e: Exception => throw new KafkaException(s"Failed to search for offset $targetOffset due to", e)
+    }
+
+    if (foundOffset < 0) null
+    else OffsetPosition(foundOffset, foundPosition)
+  }
+
+  /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   * The method will return
+   * <ul>
+   * <li>None if there is no message after the starting position.
+   * <li>TimestampOffset( largest_timestamp_checked_before_target_timestamp, the_next_offset_to_read ) otherwise.
+   * This is enssentially a time index entry built for all the messages from starting position until it finds
+   * the target timestamp or reaches the end of the segment. If the message whose timestamp is greater than
+   * or equals to the target timestamp is found, the next_offset_to_read will be the offset of that message,
+   * otherwise, it will be the next offset of the last offset in this segment.
+   * </ul>
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists after starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] ={
+    var lastCheckedOffset = -1L
+    var maxTimestampChecked = Message.NoTimestamp
+    var foundOrAborted = false
+    def handleMessage(offset: Long, position: Int, magic: Byte, timestamp: Long): Boolean = {
+      lastCheckedOffset = offset
+      // Stop searching once we see message format before 0.10.0 or we found the message.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (magic == Message.MagicValue_V0) {
+        foundOrAborted = true
+        true
+      } else if (timestamp >= targetTimestamp) {
+        foundOrAborted = true
+        true
+        // Continue searching but record the last checked offset.
+      } else {
+        maxTimestampChecked = math.max(maxTimestampChecked, timestamp)
+        false
+      }
+    }
+
+    try scanAndHandleMessages(startingPosition, handleMessage)
+    catch {
+      case e: Exception => throw new KafkaException(s"Failed to search target timestamp $targetTimestamp due to", e)
+    }
+
+    if (foundOrAborted)
+      Some(TimestampOffset(maxTimestampChecked, lastCheckedOffset))
+    else if (lastCheckedOffset >= 0)
+      // we have finished scanning the log, the offset looked for is after the last checked offset.
+      Some(TimestampOffset(maxTimestampChecked, lastCheckedOffset + 1))
```

Comment:
Hmm, if we return None, what timestamp and offset response should we return to the user? We cannot simply use the log end offset or HW because it is possible the log end offset or HW has grown after the search. So it seems we should tell the caller that up until which offset we have searched?